File: hub/webui/static/screens/bugreport.js
/* Bug report screen — OWNED BY THE BUG REPORT SCREEN.
 *
 * The only JS file this screen edits. It draws one box, one button and the line that says who the
 * report goes as, and calls the verbs hub/webui/screens/bugreport.py registers (bug_set_text,
 * bug_send). Nothing here decides anything: the session refuses an empty box, a signed-out hub
 * and a second report inside five seconds, and the server refuses the same again.
 *
 * i18n: `nav_bugreport` lives in i18n.py and is read with ctx.t(); everything else ships inside
 * the snapshot slice (state.bugreport.strings) and is read with the local st() helper — NOT t() —
 * so the front-end key-scan test stays honest and i18n.py is not edited for one screen. */
(function () {
  "use strict";

  window.HubUI.registerScreen("bugreport", { render: render });

  // The countdown under the send button, ticking between snapshots.
  //
  // WHY A LOCAL TICKER AND NOT THE SNAPSHOT. The snapshot is rebuilt when the SESSION changes, not
  // on a clock (hub/webui/panel.py on_change), so a "wait 4s" drawn from it would sit there saying
  // 4 until something else moved. The session does re-emit once the cooldown expires, which is what
  // actually re-enables the button; this only keeps the number honest in between. It is cleared at
  // the top of every render because core.render() throws the whole screen away each time.
  var tick = null;

  function stopTick() {
    if (tick !== null) { clearInterval(tick); tick = null; }
  }

  function render(root, state, ctx) {
    var ui = ctx.ui, call = ctx.call, t = ctx.t;
    var el = ui.el;
    var s = state.bugreport || {};
    var strings = s.strings || {};
    var limit = s.max || 2000;
    // THE BOX'S ID CARRIES THE SEND COUNTER, and that is load-bearing rather than tidy. The core
    // saves every #app input's value before it rebuilds the screen and writes it back afterwards
    // BY ID (core.js restoreInputs), so that a snapshot landing mid-sentence cannot wipe what
    // somebody is typing. It does that unconditionally, which means the empty box drawn after a
    // successful send was refilled with the report that had just been filed - one press of Send
    // away from filing it a second time. A filed report bumps `seq` (hub/competitive.py
    // bug_seq), the id changes, the restore finds no such element and skips it, and the box is
    // properly empty. Every other render keeps the same id and the same protection.
    var boxId = "bugreport-text-" + (s.seq || 0);

    stopTick();

    // Local string lookup for this screen's own slice (see the i18n note above).
    function st(key, kw) {
      var v = (strings && strings[key] !== undefined) ? strings[key] : key;
      if (kw) {
        v = String(v).replace(/\{(\w+)\}/g, function (m, name) {
          return (kw[name] !== undefined && kw[name] !== null) ? String(kw[name]) : m;
        });
      }
      return v;
    }

    var wrap = el("div", "bug");

    var head = el("div", "bug-head");
    head.appendChild(el("h1", null, t("nav_bugreport")));
    head.appendChild(el("p", "bug-sub", st("subtitle")));
    wrap.appendChild(head);

    var card = el("div", "bug-card");

    // -- who it goes as -------------------------------------------------------
    // Shown rather than asked for: the id and the name are taken off the signed-in account by the
    // server, so the honest thing is to say whose they are before anything is sent.
    var who = el("div", "bug-who");
    if (s.signed_in) {
      who.appendChild(ui.avatar({ text: ui.initials(s.persona || ""), ariaLabel: s.persona || "" }));
      var whoText = el("div", "bug-whotext");
      whoText.appendChild(el("span", "bug-whoname", st("as", { name: s.persona || "" })));
      whoText.appendChild(el("span", "bug-whoid", st("as_id", { id: s.steam_id || "" })));
      who.appendChild(whoText);
    } else {
      who.appendChild(el("span", "bug-whoname", st("err_signed_out")));
    }
    card.appendChild(who);

    // -- the box --------------------------------------------------------------
    var field = el("div", "bug-field");
    var label = el("label", "bug-label", st("label"));
    label.setAttribute("for", boxId);
    field.appendChild(label);

    var box = document.createElement("textarea");
    box.id = boxId;                         // see boxId above: the core preserves an input's
    box.className = "bug-box";              // value BY ID across re-renders
    box.rows = 8;
    box.maxLength = limit;
    box.placeholder = st("placeholder");
    box.value = s.text || "";
    box.disabled = !s.signed_in;
    field.appendChild(box);

    var meter = el("div", "bug-meter");
    var left = el("span", "bug-left");
    left.id = "bugreport-left";
    left.textContent = st("left", { n: Math.max(0, limit - (box.value || "").length) });
    meter.appendChild(left);
    meter.appendChild(el("span", "bug-limit", st("limit", { n: s.cooldown_seconds || 5 })));
    field.appendChild(meter);
    card.appendChild(field);

    // -- send -----------------------------------------------------------------
    var cooling = (s.cooldown_ms || 0) > 0;
    // A local deadline rather than the raw number: the ticker below counts down against it while
    // the snapshot stands still.
    var until = cooling ? (Date.now() + s.cooldown_ms) : 0;

    var actions = el("div", "bug-actions");
    var send = ui.btn("bug-send", sendLabel(), function () {
      var node = document.getElementById(boxId);
      call("bug_send", node ? node.value : "");
    }, { tag: "button" });
    send.id = "bugreport-send";
    actions.appendChild(send);

    var note = el("div", "bug-note");
    note.id = "bugreport-note";
    card.appendChild(actions);
    card.appendChild(note);
    card.appendChild(el("div", "bug-privacy", st("privacy")));
    wrap.appendChild(card);
    root.appendChild(wrap);

    // The draft is synced on `change` (which fires on blur), NOT on every keystroke: the verb is
    // an HTTP POST to the local bridge and one per character typed is a lot of traffic for a
    // convenience. Send carries the live value itself, so nothing typed can be lost by this.
    box.addEventListener("change", function () { call("bug_set_text", box.value); });
    box.addEventListener("input", function () {
      var n = document.getElementById("bugreport-left");
      if (n) { n.textContent = st("left", { n: Math.max(0, limit - box.value.length) }); }
      refresh();
    });

    refresh();
    if (cooling) { tick = setInterval(refresh, 250); }

    function secondsLeft() {
      return until ? Math.max(0, Math.ceil((until - Date.now()) / 1000)) : 0;
    }

    function sendLabel() {
      if (s.sending) { return st("sending"); }
      var n = secondsLeft();
      if (n > 0) { return st("wait", { n: n }); }
      return s.sent ? st("sent_again") : st("send");
    }

    /** Re-label and re-enable the button, and redraw the line under it, without a full render. */
    function refresh() {
      var btn = document.getElementById("bugreport-send");
      var node = document.getElementById(boxId);
      var noteEl = document.getElementById("bugreport-note");
      if (!btn) { stopTick(); return; }
      var n = secondsLeft();
      if (n === 0 && tick !== null) { stopTick(); }
      btn.textContent = sendLabel();
      var empty = !node || !String(node.value || "").trim();
      btn.disabled = Boolean(s.sending) || n > 0 || empty || !s.signed_in;
      if (!noteEl) { return; }
      noteEl.className = "bug-note";
      var text = "";
      // THE COUNTDOWN IS NOT A TELLING-OFF. It lives on the button; this line only says something
      // when there is something to say. Keying it on `n > 0` as well as on the error put "One
      // report every 5 seconds - try again in a moment" under a report that had just been
      // accepted, which reads as a refusal of the thing that in fact worked. Only a send that was
      // actually REFUSED (error === "too_fast") earns the warning.
      if (s.error === "too_fast") {
        text = st("err_too_fast", { n: s.cooldown_seconds || 5 });
        noteEl.classList.add("warn");
      } else if (s.error === "empty") {
        text = st("err_empty");
        noteEl.classList.add("warn");
      } else if (s.error === "signed_out") {
        text = st("err_signed_out");
        noteEl.classList.add("warn");
      } else if (s.error) {
        text = st("err_failed");
        noteEl.classList.add("warn");
      } else if (s.sent) {
        text = st("sent");
        noteEl.classList.add("ok");
      }
      noteEl.textContent = text;
    }
  }
})();

File: hub/webui/screens/bugreport.py
"""Bug report screen — the Python half: its snapshot slice, its bridge verbs, and its strings.

OWNED BY THE BUG REPORT SCREEN. Sam, 2026-09-16: "the settings button at the top - let's put
another button that's called Bug report, and it opens a screen that enables the user to report a
bug. That bug will then get sent to the admin console with their steam id and name. Only allow the
user to send 1 bug every 5 seconds."

WHO IT IS FROM IS NOT A FIELD. The steam id and the persona come off the signed-in account
(``session.me``), travel with the bearer token, and are stamped onto the record by the SERVER
(server/live.cjs submitBugReport) - the page never sends them. A box labelled "your Steam ID" is a
box to type somebody else's into, and the one thing a bug report has to be good for is going back
to the person who filed it. The screen shows who it will be sent as, so nobody is surprised.

THE FIVE-SECOND LIMIT IS NOT DRAWN HERE. It is enforced in the SESSION
(hub/competitive.py send_bug_report / bug_cooldown_left) and again on the server, and this slice
only reports how much of it is left so the page can say so. A limit that lives in the page is a
limit a reload clears, and there are two UIs over this one session.

Nothing here is authoritative: the verbs are 1:1 passthroughs onto the session on the UI thread,
exactly like every other screen in this package. No pywebview and no Tk, so it stays testable
headless (tests/test_screen_bugreport.py).
"""
from . import register_snapshot, register_verbs
from ... import competitive as comp
from ... import i18n


# ---------------------------------------------------------------- per-language strings
# Only the strings i18n.py does not already carry; `nav_bugreport` DOES live in i18n.py (the core
# nav draws it) and is looked up by the JS through ctx.t(). Every language below has the same key
# set, so the app's 7-language coverage is unchanged.
_S = {
    "en": {
        "subtitle": "Something broken? Tell us what happened.",
        "label": "What went wrong?",
        "placeholder": "What were you doing, and what did Lights Out do instead?",
        "send": "Send report",
        "sending": "Sending…",
        "wait": "Wait {n}s",
        "sent": "Thanks - your report is with us.",
        "sent_again": "Send another",
        "as": "Sent as {name}",
        "as_id": "Player ID {id}",
        "privacy": "We receive your player profile name and ID, report text, submission time, and app and mode versions.",
        "left": "{n} characters left",
        "limit": "One report every {n} seconds.",
        "err_empty": "Write what went wrong first.",
        "err_signed_out": "Sign in to send a bug report.",
        "err_too_fast": "One report every {n} seconds - try again in a moment.",
        "err_failed": "That could not be sent. Try again in a moment.",
    },
    "de": {
        "subtitle": "Etwas kaputt? Sag uns, was passiert ist.",
        "label": "Was ist schiefgelaufen?",
        "placeholder": "Was hast du gemacht, und was hat Lights Out stattdessen getan?",
        "send": "Melden",
        "sending": "Wird gesendet…",
        "wait": "Noch {n}s",
        "sent": "Danke - deine Meldung ist bei uns.",
        "sent_again": "Weitere senden",
        "as": "Gesendet als {name}",
        "as_id": "Spieler-ID {id}",
        "privacy": "Wir erhalten deinen Profilnamen und deine Spieler-ID, den Meldungstext, die Sendezeit sowie die App- und Spielmodusversionen.",
        "left": "Noch {n} Zeichen",
        "limit": "Eine Meldung alle {n} Sekunden.",
        "err_empty": "Schreib zuerst, was schiefgelaufen ist.",
        "err_signed_out": "Melde dich an, um einen Fehler zu melden.",
        "err_too_fast": "Eine Meldung alle {n} Sekunden - versuch es gleich noch einmal.",
        "err_failed": "Das konnte nicht gesendet werden. Versuch es gleich noch einmal.",
    },
    "es": {
        "subtitle": "¿Algo va mal? Cuéntanos qué ha pasado.",
        "label": "¿Qué ha fallado?",
        "placeholder": "¿Qué estabas haciendo y qué hizo Lights Out en su lugar?",
        "send": "Enviar informe",
        "sending": "Enviando…",
        "wait": "Espera {n}s",
        "sent": "Gracias: ya tenemos tu informe.",
        "sent_again": "Enviar otro",
        "as": "Enviado como {name}",
        "as_id": "ID de jugador {id}",
        "privacy": "Recibimos tu nombre e ID de perfil, el texto del informe, la hora de envío y las versiones de la app y del modo.",
        "left": "Quedan {n} caracteres",
        "limit": "Un informe cada {n} segundos.",
        "err_empty": "Escribe primero qué ha fallado.",
        "err_signed_out": "Inicia sesión para enviar un informe.",
        "err_too_fast": "Un informe cada {n} segundos: inténtalo en un momento.",
        "err_failed": "No se ha podido enviar. Inténtalo en un momento.",
    },
    "fr": {
        "subtitle": "Quelque chose ne va pas ? Dis-nous ce qui s'est passé.",
        "label": "Qu'est-ce qui a échoué ?",
        "placeholder": "Que faisais-tu, et qu'a fait Lights Out à la place ?",
        "send": "Envoyer",
        "sending": "Envoi…",
        "wait": "Encore {n}s",
        "sent": "Merci - ton signalement nous est parvenu.",
        "sent_again": "En envoyer un autre",
        "as": "Envoyé en tant que {name}",
        "as_id": "ID de joueur {id}",
        "privacy": "Nous recevons le nom et l’ID de ton profil, le texte du signalement, l’heure d’envoi et les versions de l’application et du mode.",
        "left": "{n} caractères restants",
        "limit": "Un signalement toutes les {n} secondes.",
        "err_empty": "Écris d'abord ce qui n'a pas marché.",
        "err_signed_out": "Connecte-toi pour signaler un bug.",
        "err_too_fast": "Un signalement toutes les {n} secondes - réessaie dans un instant.",
        "err_failed": "Impossible d'envoyer. Réessaie dans un instant.",
    },
    "pt": {
        "subtitle": "Algo avariado? Diz-nos o que aconteceu.",
        "label": "O que correu mal?",
        "placeholder": "O que estavas a fazer e o que fez o Lights Out em vez disso?",
        "send": "Enviar relatório",
        "sending": "A enviar…",
        "wait": "Espera {n}s",
        "sent": "Obrigado - o teu relatório chegou.",
        "sent_again": "Enviar outro",
        "as": "Enviado como {name}",
        "as_id": "ID de jogador {id}",
        "privacy": "Recebemos o nome e ID do teu perfil, o texto do relatório, a hora de envio e as versões da app e do modo.",
        "left": "Faltam {n} caracteres",
        "limit": "Um relatório a cada {n} segundos.",
        "err_empty": "Escreve primeiro o que correu mal.",
        "err_signed_out": "Inicia sessão para enviar um relatório.",
        "err_too_fast": "Um relatório a cada {n} segundos - tenta daqui a pouco.",
        "err_failed": "Não foi possível enviar. Tenta daqui a pouco.",
    },
    "ru": {
        "subtitle": "Что-то сломалось? Расскажите, что случилось.",
        "label": "Что пошло не так?",
        "placeholder": "Что вы делали и что вместо этого сделал Lights Out?",
        "send": "Отправить",
        "sending": "Отправка…",
        "wait": "Ещё {n} с",
        "sent": "Спасибо - отчёт получен.",
        "sent_again": "Отправить ещё",
        "as": "Отправлено от имени {name}",
        "as_id": "ID игрока {id}",
        "privacy": "Мы получаем имя и ID вашего профиля, текст отчёта, время отправки, версии приложения и режима.",
        "left": "Осталось символов: {n}",
        "limit": "Один отчёт раз в {n} с.",
        "err_empty": "Сначала напишите, что пошло не так.",
        "err_signed_out": "Войдите, чтобы отправить отчёт.",
        "err_too_fast": "Один отчёт раз в {n} с - попробуйте через мгновение.",
        "err_failed": "Отправить не удалось. Попробуйте через мгновение.",
    },
    "zh": {
        "subtitle": "出问题了？告诉我们发生了什么。",
        "label": "哪里出错了？",
        "placeholder": "你当时在做什么，Lights Out 又做了什么？",
        "send": "发送报告",
        "sending": "发送中…",
        "wait": "还需 {n} 秒",
        "sent": "谢谢，我们已收到你的报告。",
        "sent_again": "再发一条",
        "as": "以 {name} 的身份发送",
        "as_id": "玩家 ID {id}",
        "privacy": "我们会收到你的玩家资料名称和 ID、报告内容、提交时间以及应用和模式版本。",
        "left": "还可输入 {n} 个字符",
        "limit": "每 {n} 秒只能发送一条。",
        "err_empty": "请先写下出了什么问题。",
        "err_signed_out": "请先登录再发送报告。",
        "err_too_fast": "每 {n} 秒只能发送一条 - 请稍候再试。",
        "err_failed": "发送失败，请稍候再试。",
    },
}


def _strings(lang: str) -> dict:
    """This screen's strings for the active language, English-filled so a missing key can never
    render as the raw key (the same English-fallback rule i18n.py uses)."""
    merged = dict(_S.get(i18n.DEFAULT, {}))
    merged.update(_S.get(lang, {}))
    return merged


def _cooldown_ms(session) -> int:
    """How much of the five seconds is left, in whole milliseconds.

    Read from the SESSION, never recomputed here: hub/competitive.py owns the rule and the server
    owns the last word on it."""
    left = getattr(session, "bug_cooldown_left", None)
    try:
        return int(max(0.0, float(left())) * 1000) if callable(left) else 0
    except Exception:                    # noqa: BLE001 — a clock must never break the snapshot
        return 0


@register_snapshot("bugreport")
def snapshot(session, panel) -> dict:
    """The ``bugreport`` slice: who the report goes as, what is in the box, and where the send
    stands (in flight / filed / refused, and how long the cooldown has left to run)."""
    me = getattr(session, "me", None) or {}
    return {
        "bugreport": {
            "signed_in": bool(me),
            "persona": me.get("name") or "",
            "steam_id": me.get("steam_id") or "",
            # The draft lives on the session rather than in the page, so switching to Settings and
            # back does not throw away what somebody has half-written.
            "text": str(getattr(session, "bug_text", "") or ""),
            # Which box this is. It changes only when a report is FILED, and the page hangs the
            # textarea's id off it so the filed text cannot be restored into the empty box.
            "seq": int(getattr(session, "bug_seq", 0) or 0),
            "max": comp.BUG_TEXT_MAX,
            "sending": bool(getattr(session, "bug_sending", False)),
            "sent": bool(getattr(session, "bug_sent", False)),
            # "" | empty | signed_out | too_fast | failed | whatever the server called it
            "error": str(getattr(session, "bug_error", "") or ""),
            "cooldown_ms": _cooldown_ms(session),
            "cooldown_seconds": int(comp.BUG_COOLDOWN_SECONDS),
            "strings": _strings(i18n.get_language()),
        }
    }


# ---------------------------------------------------------------- verb implementations
def _set_text(panel, text=""):
    """Keep the draft. Called on blur/change rather than per keystroke - see the JS for why."""
    def apply():
        setter = getattr(panel.session, "set_bug_text", None)
        if callable(setter):
            setter(text)
    panel.post(apply)


def _send(panel, text=""):
    """Send it. Every guard (empty, signed out, too soon) lives in the session method."""
    def apply():
        send = getattr(panel.session, "send_bug_report", None)
        if callable(send):
            send(text)
    panel.post(apply)


# Verb names are screen-scoped so they cannot collide with the core verbs or another screen's.
register_verbs("bugreport", {
    "bug_set_text": _set_text,
    "bug_send":     _send,
})
